File: projects/28-telecom-tower-monitoring/fog/gateway.js
import process from "node:process";
import { fileURLToPath } from "node:url";
import express from "express";
import { Windower } from "./windower.js";
import { Dispatcher } from "./dispatcher.js";
import { evaluate, thresholds } from "./alarms.js";

function validate(body) {
  if (!body || typeof body !== "object" || Array.isArray(body)) return "body must be a JSON object";
  const { sensor_type, site_id, readings } = body;
  if (typeof sensor_type !== "string" || sensor_type.length === 0) return "sensor_type is required";
  if (typeof site_id !== "string" || site_id.length === 0) return "site_id is required";
  if (!Array.isArray(readings) || readings.length === 0) return "readings must be a non-empty array";
  for (const r of readings) {
    if (!r || typeof r.value !== "number" || Number.isNaN(r.value)) return "each reading needs a numeric value";
  }
  return null;
}

function enrich(batch) {
  return batch.map((w) => ({ ...w, alerts: evaluate(w) }));
}

function buildApp(options = {}) {
  const windowMs = options.windowMs ?? Number(process.env.WINDOW_SECONDS || 10) * 1000;
  const windower = options.windower ?? new Windower(windowMs);
  const sink = options.sink ?? (async () => {});

  windower.on("flush", (batch) => {
    Promise.resolve(sink(enrich(batch))).catch((err) => console.error(`sink failed: ${err.message}`));
  });

  const app = express();
  app.use(express.json({ limit: "1mb" }));
  app.locals.windower = windower;

  app.post("/ingest", (req, res) => {
    const problem = validate(req.body);
    if (problem) return res.status(400).json({ error: problem });
    const { sensor_type, site_id, unit } = req.body;
    const now = new Date().toISOString();
    for (const r of req.body.readings) {
      windower.accept({ sensor_type, site_id, unit, value: r.value, ts: r.ts || now });
    }
    return res.status(202).json({ accepted: req.body.readings.length });
  });

  app.get("/health", (req, res) => res.json({ status: "ok", pending: windower.pending() }));
  app.get("/thresholds", (req, res) => res.json(thresholds()));

  return app;
}

async function main() {
  const port = Number(process.env.PORT || 8000);
  const dispatcher = new Dispatcher();
  await dispatcher.configure();
  const app = buildApp({ sink: (windows) => dispatcher.publish(windows) });
  app.locals.windower.start();
  app.listen(port, () => console.log(`tower gateway listening on :${port}`));
}

const isEntry = process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1];
if (isEntry) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

export { buildApp, validate, enrich };

File: projects/27-river-flood-monitoring/fog/gateway.py
"""Fog node on aiohttp: async ingest into a lock-free per-group buffer, background window flush to SQS."""
import asyncio
import os
from datetime import datetime, timedelta, timezone

from aiohttp import web

import publisher
from staging import evaluate, thresholds_payload
from windowing import summarise

WINDOW_SECONDS = float(os.getenv("WINDOW_SECONDS", "10"))
QUEUE_NAME = os.getenv("SQS_QUEUE_NAME", "rfw-catchment-agg")
ENDPOINT = os.getenv("AWS_ENDPOINT_URL")
REGION = os.getenv("AWS_REGION", "eu-west-1")

REQUIRED = ("sensor_type", "site_id", "readings")


def validate(body):
    if not isinstance(body, dict):
        return "body must be an object"
    for field in REQUIRED:
        if field not in body:
            return f"missing field: {field}"
    if not isinstance(body["readings"], list) or not body["readings"]:
        return "readings must be a non-empty list"
    for reading in body["readings"]:
        if not isinstance(reading, dict) or not isinstance(reading.get("value"), (int, float)):
            return "each reading needs a numeric value"
    return None


async def ingest(request):
    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "invalid JSON"}, status=400)
    error = validate(body)
    if error:
        return web.json_response({"error": error}, status=400)
    key = (body["sensor_type"], body["site_id"])
    request.app["buffers"].setdefault(key, []).extend(body["readings"])
    request.app["units"][body["sensor_type"]] = body.get("unit", "")
    return web.json_response({"accepted": len(body["readings"])}, status=202)


async def health(request):
    return web.json_response({"status": "ok"})


async def thresholds(request):
    return web.json_response(thresholds_payload())


def drain(buffers):
    snapshot = {key: values for key, values in buffers.items() if values}
    buffers.clear()
    return snapshot


async def flush_loop(app):
    while True:
        await asyncio.sleep(WINDOW_SECONDS)
        snapshot = drain(app["buffers"])
        if not snapshot:
            continue
        end = datetime.now(timezone.utc)
        start = end - timedelta(seconds=WINDOW_SECONDS)
        messages = []
        for (sensor_type, site_id), readings in snapshot.items():
            agg = summarise(sensor_type, site_id, app["units"].get(sensor_type, ""),
                            readings, start.isoformat(), end.isoformat(), WINDOW_SECONDS)
            agg["alerts"] = evaluate(sensor_type, agg)
            messages.append(agg)
        try:
            await asyncio.to_thread(publisher.send_window, messages)
        except Exception as exc:
            print(f"flush failed: {exc}", flush=True)


async def on_start(app):
    await asyncio.to_thread(publisher.configure, ENDPOINT, REGION, QUEUE_NAME)
    app["flusher"] = asyncio.create_task(flush_loop(app))


def build_app():
    app = web.Application()
    app["buffers"] = {}
    app["units"] = {}
    app.add_routes([
        web.post("/ingest", ingest),
        web.get("/health", health),
        web.get("/thresholds", thresholds),
    ])
    return app


def main():
    app = build_app()
    app.on_startup.append(on_start)
    print(f"fog listening on :8000 (window={WINDOW_SECONDS}s)", flush=True)
    web.run_app(app, host="0.0.0.0", port=8000, print=None)


if __name__ == "__main__":
    main()
